### ISSUE TITLE
BatchNormalization doesn't work in test mode sometimes

### ISSUE THREAD

**Issue by szagoruyko** [torch/nn#308]
I have observed this issue on several architectures for classification. One time training error goes down, test error stales on one class, next time I rerun the training process and it's fine, both go down.

**Comment by soumith** [comment]
hmm, are you noting that there might be a coding bug, or that because of the distribution of training samples, batchnorm's running_mean and running_std might just be bad estimates sometimes.

**Comment by soumith** [comment]
without a repro, cant do much here.

**Comment by szagoruyko** [comment]
actually I can prepare a repro test on cifar.torch. this bug annoys me a lot.

**Comment by nkoumchatzky** [comment]
@szagoruyko How do you sample your training data? Do you use the same training batch order (and same random seed) from one training process to another?

**Comment by szagoruyko** [comment]
@nkoumchatzky yes, when I encounter this problem I have to reset the seed until batchnorm starts working in test mode correctly.

**Comment by nkoumchatzky** [comment]
Ok I see what you mean. So as @soumith said, this is probably not be a bug but more likely the built-in B.N. behavior due to these running_mean and running_std estimates' uncertainty from one run to another (provided you change the seed or the architecture).
